salt/states/snap.py

```diff
@@ -0,0 +1,125 @@
+# -*- coding: utf-8 -*-
+"""
+Management of snap packages
+===============================================
+"""
+from __future__ import absolute_import, print_function, unicode_literals
+
+import salt.utils.path
+
+__virtualname__ = "snap"
+
+
+def __virtual__():
+    if salt.utils.path.which("snap"):
+        return __virtualname__
+
+    return (
+        False,
+        'The snap state module cannot be loaded: the "snap" binary is not in the path.',
+    )
+
+
+def installed(name, channel=None):
+    """
+    Ensure that the named snap package is installed
+
+    name
+        The snap package
+
+    channel
+        Optional. The channel to install the package from.
+    """
+    ret = {"name": name, "changes": {}, "pchanges": {}, "result": None, "comment": ""}
+
+    old = __salt__["snap.versions_installed"](name)
+    if not old:
+        if __opts__["test"]:
+            ret["comment"] = 'Package "{0}" would have been installed'.format(name)
+            ret["pchanges"]["new"] = name
+            ret["pchanges"]["old"] = None
+            ret["result"] = None
+            return ret
+
+        install = __salt__["snap.install"](name, channel=channel)
+        if install["result"]:
+            ret["comment"] = 'Package "{0}" was installed'.format(name)
+            ret["changes"]["new"] = name
+            ret["changes"]["old"] = None
+            ret["result"] = True
+            return ret
+
+        ret["comment"] = 'Package "{0}" failed to install'.format(name)
+        ret["comment"] += "\noutput:\n" + install["output"]
+        ret["result"] = False
+        return ret
+
+    # Currently snap always returns only one line?
+    old_channel = old[0]["tracking"]
+    if old_channel != channel and channel is not None:
+        if __opts__["test"]:
+            ret[
+                "comment"
+            ] = 'Package "{0}" would have been switched to channel {1}'.format(
+                name, channel
+            )
+            ret["pchanges"]["old_channel"] = old_channel
+            ret["pchanges"]["new_channel"] = channel
+            ret["result"] = None
+            return ret
+
+        refresh = __salt__["snap.install"](name, channel=channel, refresh=True)
+        if refresh["result"]:
+            ret["comment"] = 'Package "{0}" was switched to channel {1}'.format(
+                name, channel
+            )
+            ret["pchanges"]["old_channel"] = old_channel
+            ret["pchanges"]["new_channel"] = channel
+            ret["result"] = True
+            return ret
+
+        ret["comment"] = 'Failed to switch Package "{0}" to channel {1}'.format(
+            name, channel
+        )
+        ret["comment"] += "\noutput:\n" + install["output"]
```

Comment:
I've been testing out using these modules in the [`packages-formula`](https://github.com/saltstack-formulas/packages-formula), to replace the use of `cmd.run` for `snap` installations.  Hit a traceback with this line:

```python-traceback
An exception occurred in this state: Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/salt/state.py", line 2179, in call
    ret = self.states[cdata["full"]](
  File "/usr/lib/python3/dist-packages/salt/loader/lazy.py", line 149, in __call__
    return self.loader.run(run_func, *args, **kwargs)
  File "/usr/lib/python3/dist-packages/salt/loader/lazy.py", line 1201, in run
    return self._last_context.run(self._run_as, _func_or_method, *args, **kwargs)
  File "/usr/lib/python3/dist-packages/salt/loader/lazy.py", line 1216, in _run_as
    return _func_or_method(*args, **kwargs)
  File "/usr/lib/python3/dist-packages/salt/loader/lazy.py", line 1249, in wrapper
    return f(*args, **kwargs)
  File "/var/cache/salt/minion/extmods/states/snap.py", line 84, in installed
    ret["comment"] += "\noutput:\n" + install["output"]
UnboundLocalError: local variable 'install' referenced before assignment
```

Looking at the code, realised that this line should actually be:

```suggestion
        ret["comment"] += "\noutput:\n" + refresh["output"]
```

Left over from a copy-paste of [L53](https://github.com/saltstack/salt/pull/56953/files#diff-f3821109a054fffd58b5182fb6b0707a5396be80c55a044150942f488619a4e7R53) above, perhaps.

@lordcirth As the original PR author, would you mind confirming this suggestion?

Reply:
@myii  Sounds correct to me

Reply:
Appreciate the response, @lordcirth.  I've got another question, if you don't mind.

I've found that the execution module functions are returning bytes instead of string objects (where relevant).  Should `__salt__["cmd.run_all"]()` be used instead of `subprocess.check_output()`?  Checking across other modules, that seems to be the main method being used when working with command output.

Reply:
I'm not very experienced with Python or Saltstack internals. Your guess is probably better than mine.

Reply:
@lordcirth OK, If I get around to pushing this to the `packages-formula`, I'll try to refactor it accordingly.  I'll report back here on how things work out.  Thanks again for the prompt response -- and for providing these modules in the first place!